File: luogu/P1083-binarysearch.py
###
# Solution: binary search + difference array, TLE for python code
# to quickly hold for the interval modify (each minus d classrooms for day x to y
# we can use difference array, that is d[i] = a[i] - a[i-1]
# then a[i] = \sum_{j=0toi} d[i]
# so for the first k modifys, we can take O(n + k) to calculate for the a[i]
# then, we find that the remaining classroom is monotonically decreasing.
# so we can use binary search to find the first unsatisfy order,
# Complexity is O((m+n)*log m)
###

def check(arr, data, k): # check [0, k] is possible?
    deltas = [0 for _ in range(len(arr)+1)] # d[i] = arr'[i] - arr'[i-1]
    for d, x, y in data[:k+1]:
        deltas[x] += d
        deltas[y+1] -= d

    _sum = 0
    for i, num in enumerate(arr):
        _sum += deltas[i]
        if _sum > num:
            return False
    return True

def binary_search(arr, data, low, high):
    if low >= high:
        if check(arr, data, low):
            return low + 1
        else:
            return low
    mid = (low + high) >> 1
    if check(arr, data, mid):
        return binary_search(arr, data, mid+1, high)
    else:
        return binary_search(arr, data, low, mid-1)



n, m = input().split()
n, m = int(n), int(m)
arr = input().split()
arr = [int(i) for i in arr]

orders = []
for _ in range(m):
    data = input().split()
    data = [int(__) for __ in data]
    data[1] -= 1
    data[2] -= 1
    orders.append(data)

order_select = binary_search(arr, orders, 0, m-1)
if order_select == m:
    print(0)
else:
    print(f'-1\n{order_select + 1}')

File: luogu/P1083-binarysearch.cpp
/*
Solution can be find in P1083.py, note that you need to store the classroom num by long long rather than int!
*/

#include <iostream>
using namespace std;
struct Order{
    int start, end;
    long long num;
    void set_value(long long n, int s, int e) {
        this->num = n;
        this->start = s;
        this->end = e;
    }
};
bool check(long long* arr, Order** orders, int k, int n, int m) { // check [0, k] is possible?
    long long* deltas = new long long [n+1];
    for (int i = 0; i <= n; i++) {
        deltas[i] = 0ll;
    }
    for (int i = 0; i <= k; i++) {
        deltas[orders[i]->start] += orders[i]->num;
        deltas[orders[i]->end + 1] -= orders[i]->num;
    }
    long long _sum = 0;
    for (int i = 0; i < n; i++) {
        _sum += deltas[i];
        if (_sum > arr[i]) {
            delete []deltas;
            return false;
        }
    }
    delete []deltas;
    return true;
}

int binary_search(long long* arr, Order** orders, int low, int high, int n, int m) {
    if (low >= high) {
        if (check(arr, orders, low, n, m)) {
            return low + 1;
        } else {
            return low;
        }
    }
    int mid = (low + high) >> 1;
    if (check(arr, orders, mid, n, m)) {
        return binary_search(arr, orders, mid+1, high, n, m);
    } else {
        return binary_search(arr, orders, low, mid-1, n, m);
    }
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    int n, m;
    scanf("%d%d", &n, &m);
    long long* arr = new long long [n];
    for (int i = 0; i < n; i++) {
        scanf("%lld", &arr[i]);
    }
    Order** orders = new Order* [m];
    int s, e;
    long long d;
    for (int i = 0; i < m; i++) {
        orders[i] = new Order();
        scanf("%lld%d%d", &d, &s, &e);
        orders[i]->set_value(d, s-1, e-1);
    }

    int order_select = binary_search(arr, orders, 0, m-1, n, m);
    if (order_select == m) {
        printf("0\n");
    } else {
        printf("-1\n%d\n", order_select + 1);
    }





    delete []orders;
    delete []arr;
    return 0;
}